components/tooltip/tooltip.js

```diff
@@ -0,0 +1,579 @@
+		return this.shadowRoot.querySelector('.d2l-tooltip-content');
+	}
+
+	__getTooltipTarget() {
+		return this.shadowRoot.querySelector('.d2l-tooltip-target-position');
+	}
+
+	async _layoutTooltip() {
+
+		const target = this._target;
+		if (!target) {
+			return;
+		}
+		const tooltipTarget = this.__getTooltipTarget();
+		if (!tooltipTarget) {
+			return;
+		}
+
+		const targetRect =  target.getBoundingClientRect();
+		const spaceAround = this._constrainSpaceAround({
+			above: targetRect.top - this._viewportMargin,
+			below: window.innerHeight - (targetRect.top + targetRect.height) - this._viewportMargin,
+			left: targetRect.left - this._viewportMargin,
+			right: document.documentElement.clientWidth - (targetRect.left + targetRect.width) - this._viewportMargin
+		});
+
+		const verticalWidth = Math.max(spaceAround.left + targetRect.width + spaceAround.right, 0);
+		const horizontalHeight = Math.max(spaceAround.above + targetRect.height + spaceAround.below, 0);
+
+		const spaces = [
+			{ dir: 'top', width: verticalWidth, height: Math.max(spaceAround.above - this.offset, 0) },
+			{ dir: 'bottom', width: verticalWidth, height: Math.max(spaceAround.below - this.offset, 0) },
+			{ dir: 'right', width: Math.max(spaceAround.right - this.offset, 0), height: horizontalHeight },
+			{ dir: 'left', width: Math.max(spaceAround.left - this.offset, 0), height: horizontalHeight }
+		];
+		if (this.getAttribute('dir') === 'rtl') {
+			const tmp = spaces[2];
+			spaces[2] = spaces[3];
+			spaces[3] = tmp;
+		}
+
+		let space = null;
+		const content = this.__getContentContainer();
+		if (this.position !== undefined) {
+			space = spaces.filter(space => space.dir === this.position)[0];
+			this._maxWidth = space.width;
+			this._maxHeight = null;
+			await this.updateComplete;
+		} else {
+			for (let i = 0; i < spaces.length; ++i) {
+				if (await this._canFitSpace(content, spaces[i])) {
+					space = spaces[i];
+					break;
+				}
+			}
+			if (space === null) {
+				space = this._findLargestSpace(spaces);
+				this._maxWidth = space.width;
+				this._maxHeight = null;
+				await this.updateComplete;
+			}
+		}
+		const contentRect = content.getBoundingClientRect();
+		this._maxWidth = Math.max(contentRect.width, content.scrollWidth);
+		this._maxHeight = contentRect.height;
+		this.openDir = space.dir;
+
+		const tooltipRect = tooltipTarget.getBoundingClientRect();
+
+		const isVertical = this._isAboveOrBelow();
+		let spaceLeft, spaceRight, centerDelta;
+		if (isVertical) {
+			const isRtl = this.getAttribute('dir') === 'rtl';
+			spaceLeft = !isRtl ? spaceAround.left : spaceAround.right;
+			spaceRight = !isRtl ? spaceAround.right : spaceAround.left;
+			centerDelta = contentRect.width - targetRect.width;
+		} else {
+			spaceLeft = spaceAround.above;
+			spaceRight = spaceAround.below;
+			centerDelta = contentRect.height - targetRect.height;
+		}
+		this._tooltipShift = this._computeTooltipShift(centerDelta, spaceLeft, spaceRight);
```

Comment:
Does this even need to be a private variable or could it just be a local `const`?